### DISCUSSION TITLE
Alternative Search to Algolia

### DISCUSSION THREAD

**Discussion by cptmedic** [Lednerb/bilberry-hugo-theme#552] [Q&A]
Good morning/afternoon all,

Is there anyway to offer searching without having to use Algolia? My main concern is eventually needing to subscribe to one of their paid plans. Is there any way to locally implement lunr.js?

Thank you!

**Comment by igor-baiborodine** [comment]
@cptmedic, it shouldn't be difficult to implement a search functionality using [lunr.js](https://lunrjs.com/).

Usually, when your site is deployed, the `index.json` file is generated and available under the website's root, for example, Bilberry Sandbox's [index file](https://www.bilberry-sandbox.kiroule.com/index.json).

You'll have to re-implement the [search part](https://github.com/Lednerb/bilberry-hugo-theme/blob/3b759939f6953b1d6cab54ffdd7a9d6699259e5a/assets/js/theme.js#L157) in the `assets/js/theme.js` file where you search corresponding indexes in the `index.json` file and display them in the search field's dropdown.

**Reply by cptmedic** [reply]
Awesome! And from your experience, have you ever came close to reaching the free tier limit from the search provider that comes with baked in support?  Just trying to evaluate future need.

**Reply by igor-baiborodine** [reply]
Here is the latest weekly usage report for my blog at kiroule.com:

![image](https://github.com/Lednerb/bilberry-hugo-theme/assets/1027701/99f16e91-e4e0-44fd-80e6-9e1e1d109f90)

As you can see, there were only four searches on my website for the current billing period. Let's say to reach the quota of 50K operations, you will need 10K visitors per month with 5 searches per visitor. Then I uploaded 386 records(indexes) so far for this month. My website has, for now, around 30 articles, and every time I publish a new article, all existing indexes are deleted, and new indexes for all articles are uploaded. But you can opt for uploading indexes only for new or modified content, for example, see this Algolia [documentation](https://www.algolia.com/doc/api-reference/api-methods/partial-update-objects/).

**Comment by Magneticdud** [comment]
I found an alternative to algolia, made in JavaScript so it will be static and work in browser: [flexsearch](https://github.com/nextapps-de/flexsearch). But because I am a lousy coder, i have no idea how it works, but it's implemented in this Hugo theme: https://github.com/frjo/hugo-theme-zen/tree/main#search
